### ISSUE TITLE
Exception on income statement; TypeError: keys must be str, int, float, bool or None, not type

### ISSUE THREAD

**Issue by TRSx80** [beancount/fava#1580]
I had just installed fava 1.24.3 a few minutes ago via pipx (installed using Python 3.11.2 in a venv) on Debian Bookworm (current Testing) and got the error in the title.  Full error below.

I'm lost in Python stuff, sorry I can be more help.  Happy to provide any additional info, test, or whatever I can of course.

Thanks for all your work on `fava`, it's great and probably one of main reasons I started using Beancount!  :)

```
$ fava user.bc

Starting Fava on http://127.0.0.1:5000
Exception on /beancount/income_statement/ [GET]
Traceback (most recent call last):
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/fava/application.py", line 359, in report
    return render_template("_layout.html", active_page=report_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/flask/templating.py", line 147, in render_template
    return _render(app, template, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/flask/templating.py", line 130, in _render
    rv = template.render(context)
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/fava/templates/_layout.html", line 44, in top-level template code
    <script type="application/json" id="ledger-data">{{ get_ledger_data()|tojson }}</script>
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/jinja2/filters.py", line 1688, in do_tojson
    return htmlsafe_json_dumps(value, dumps=dumps, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/jinja2/utils.py", line 658, in htmlsafe_json_dumps
    dumps(obj, **kwargs)
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/fava/core/charts.py", line 103, in dumps
    return ENCODER.encode(obj)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/simplejson/encoder.py", line 297, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/pipx/venvs/fava/lib/python3.11/site-packages/simplejson/encoder.py", line 379, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
TypeError: keys must be str, int, float, bool or None, not type
```

**Comment by yagebu** [comment]
Hi, thanks for the report! With these JSON serialisation it is often hard to know which object/attribute is at fault...

I've tried something that might fix it in #1582 (which I just merged into main). Could you try installing Fava from main and checking whether that fixed it? If not, could you try to provide an example file that reproduces this?

**Comment by TRSx80** [comment]
I installed all dependencies and built from latest souce:

```
$ fava --version

fava, version 1.24.4.dev2+g35517388
```

And still having the issue.

> If not, could you try to provide an example file that reproduces this?

I was able to bisect my files down, and then remove personal details, until I found a reliable, minimal test case:

```
2023-04-02 open  Assets:Cash
2023-04-02 open  Expenses:Food:Groceries
2023-04-02 open  Expenses:Sundries

2018-07-07 * "a" "b"
  Assets:Cash                               -14.00 USD
  Expenses:Food:Groceries                 4.14
  Expenses:Sundries
```
I played around a while, and I may have not covered every single case, however here is what I noticed:

-  It seems to require both the 2018 date AND the line without a currency
   - In other words, if you comment out the line with 4.14 (and no currency), no crash.
   - OR, if you leave that as-is, and change '2018-07-07' to today's date (2023-04-03) for example, no crash, either.
- Further:
  - If you change account opening dates to 2018, no more crash.
  - If you add currency (USD) to the '4.14' line, no more crash.

I have no idea what that's about, and I am really curious whether you will be able to reproduce this on your end or not.

For my part, once I added a currency (USD) to the line with 4.14, the crash went away.  I even tried the last released version of Fava from pipx from before your patch (v1.24.3) and that worked fine, too.

Of course some of those are incorrect syntax, but I don't think they should completely crash Fava.
